File: web/server/spearvm_sim/protocol.py
"""Protocole binaire des frames de simulation (serveur -> navigateur).

Un frame = en-tete JSON longueur-prefixee + charge utile binaire alignee 4o :

    u32  headerLen      (little endian, longueur du JSON *sans* padding)
    u8[] headerJson     (utf-8, puis padding 0 jusqu'au multiple de 4)
    u8[] payload        (Float32Array ou Int16Array selon header["dtype"])

L'en-tete porte toujours : `kind`, `tick`, `time`, `compute_ms`, `dtype`,
`shape`, `scale` (facteur de dequantification) et `stats` (libre).
Cette forme evite un schema binaire rigide tout en gardant le cout CPU/octet
d'un flux typed-array pur cote client.
"""

from __future__ import annotations

import json
import struct
from typing import Any, Iterable

import numpy as np

MAGIC = "spearvm.sim.v1"


def quantize_i16(data: np.ndarray, amplitude: float | None = None) -> tuple[np.ndarray, float]:
    """Quantifie en int16 symetrique. Retourne (donnees, scale) avec
    valeur_reelle = donnees * scale. Divise par deux la bande passante."""
    arr = np.asarray(data, dtype=np.float32)
    if amplitude is None:
        amplitude = float(np.max(np.abs(arr))) if arr.size else 0.0
    if not np.isfinite(amplitude) or amplitude <= 1e-12:
        return np.zeros(arr.shape, dtype=np.int16), 1.0
    scale = amplitude / 32767.0
    out = np.clip(np.rint(arr / scale), -32767, 32767).astype(np.int16)
    return out, scale


def encode_frame(
    kind: str,
    payload: np.ndarray | None,
    *,
    tick: int,
    sim_time: float,
    compute_ms: float,
    shape: Iterable[int] = (),
    scale: float = 1.0,
    stats: dict[str, Any] | None = None,
) -> bytes:
    """Serialise un frame. `payload` doit etre contigu float32 ou int16."""
    if payload is None:
        buf = b""
        dtype = "none"
    else:
        arr = np.ascontiguousarray(payload)
        if arr.dtype == np.float32:
            dtype = "f32"
        elif arr.dtype == np.int16:
            dtype = "i16"
        else:  # normalisation defensive
            arr = np.ascontiguousarray(arr, dtype=np.float32)
            dtype = "f32"
        buf = arr.tobytes()

    header = {
        "magic": MAGIC,
        "kind": kind,
        "tick": tick,
        "time": round(float(sim_time), 4),
        "compute_ms": round(float(compute_ms), 3),
        "dtype": dtype,
        "shape": list(shape),
        "scale": float(scale),
        "stats": stats or {},
    }
    raw = json.dumps(header, separators=(",", ":")).encode("utf-8")
    pad = (-len(raw)) % 4
    return struct.pack("<I", len(raw)) + raw + b"\x00" * pad + buf


def decode_frame(blob: bytes) -> tuple[dict[str, Any], np.ndarray | None]:
    """Decodeur miroir — utilise par les tests (et utile aux clients Python)."""
    (hlen,) = struct.unpack_from("<I", blob, 0)
    header = json.loads(blob[4 : 4 + hlen].decode("utf-8"))
    start = 4 + hlen + ((-hlen) % 4)
    body = blob[start:]
    if header["dtype"] == "none":
        return header, None
    dt = np.float32 if header["dtype"] == "f32" else np.int16
    arr = np.frombuffer(body, dtype=dt)
    shape = tuple(header.get("shape") or ())
    if shape and int(np.prod(shape)) == arr.size:
        arr = arr.reshape(shape)
    return header, arr

File: web/client/src/net/protocol.ts
/**
 * Decodeur du protocole binaire `spearvm.sim.v1`.
 *
 *   u32 headerLen | JSON utf-8 (+ padding 4o) | payload (f32 | i16)
 *
 * Les charges i16 sont dequantifiees (`value * scale`) dans un Float32Array
 * recycle par taille : a 30 Hz on evite ainsi ~1.5 Mo/s d'allocations.
 */

export interface FrameHeader {
  magic: string;
  kind: string;
  tick: number;
  time: number;
  compute_ms: number;
  dtype: 'f32' | 'i16' | 'none';
  shape: number[];
  scale: number;
  stats: Record<string, unknown>;
}

export interface SimFrame {
  header: FrameHeader;
  /** Donnees dequantifiees. Attention : buffer recycle, ne pas conserver. */
  data: Float32Array | null;
  bytes: number;
}

const pool = new Map<number, Float32Array>();

function scratch(size: number): Float32Array {
  let buf = pool.get(size);
  if (!buf) {
    buf = new Float32Array(size);
    pool.set(size, buf);
  }
  return buf;
}

const decoder = new TextDecoder();

export function decodeFrame(buffer: ArrayBuffer): SimFrame {
  const view = new DataView(buffer);
  const headerLen = view.getUint32(0, true);
  const header = JSON.parse(
    decoder.decode(new Uint8Array(buffer, 4, headerLen)),
  ) as FrameHeader;

  const start = 4 + headerLen + ((4 - (headerLen % 4)) % 4);
  let data: Float32Array | null = null;

  if (header.dtype === 'f32') {
    const count = (buffer.byteLength - start) >> 2;
    data = scratch(count);
    data.set(new Float32Array(buffer, start, count));
  } else if (header.dtype === 'i16') {
    const count = (buffer.byteLength - start) >> 1;
    const src = new Int16Array(buffer, start, count);
    data = scratch(count);
    const s = header.scale;
    for (let i = 0; i < count; i += 1) data[i] = src[i] * s;
  }

  return { header, data, bytes: buffer.byteLength };
}

export function statNumber(stats: Record<string, unknown>, key: string, fallback = 0): number {
  const v = stats[key];
  return typeof v === 'number' && Number.isFinite(v) ? v : fallback;
}

export function statString(stats: Record<string, unknown>, key: string, fallback = '—'): string {
  const v = stats[key];
  return typeof v === 'string' ? v : fallback;
}
